### ISSUE TITLE
Build fails on new vector version

### ISSUE THREAD

**Issue by identicalsnowflake** [mgsloan/store#148]
`vector-0.12.1.2` was recently released and apparently fixes a number of longstanding bugs. However, `store` fails to build (possibly due to breaking upstream changes?) with this version with the following error:

```
src/Data/Store/Internal.hs:779:3: error:
    • Expected kind ‘* -> *’, but ‘f’ has kind ‘*’
    • In the first argument of ‘Alt’, namely ‘f’
      In the first argument of ‘Data.Vector.Unboxed.Base.Vector’, namely
        ‘(Alt f a)’
      In the first argument of ‘Store’, namely
        ‘(Data.Vector.Unboxed.Base.Vector (Alt f a))’
    |
779 | $(deriveManyStoreUnboxVector)
    |   ^^^^^^^^^^^^^^^^^^^^^^^^^^
```

**Comment by sjakobi** [comment]
@cartazio Any idea which change in `vector` might have triggered this?

**Comment by cartazio** [comment]
Eep. There shouldn’t be any such change.  What is the th doing here?  I’m afk atm and headed to day job.  But I’ll try to carve out some time to look. 

Things that’ll help me poke at this quickly 
1) what’s the definition of this th code?
2) what’s the equivalent code that was generated for 0.12.0 series vector (the src that’s generated and gets type checked ?)
3) what’s the equivalent code that’s generated now that fails to type check?

One possibility is that the new instances in unbox  that vector now provides conflicts with previously orphaned instances ?
There were a bunch of monoid ish newtypes in base that didn’t have instances for unboxed vectors previously.  Could the issue lie with thst ?

**Comment by cartazio** [comment]
https://hackage.haskell.org/package/vector-0.12.1.2/docs/Data-Vector-Unboxed.html#t:Vector

Unboxed Vector now has an instance for Alt.  

Is your unboxed instance different ? Or just an Orphan?

**Comment by cartazio** [comment]
Naively : orphan instances ?

**Comment by Shimuuar** [comment]
No TH can't work with type parameters kind other than `*`.  `store` uses TH to generate Store instances for all types which are also has `Unboxed` instance. Here is instance head that is generated for `Alt f a`:

```haskell
instance ( Store (U.Vector f)
         , Store (U.Vector a)
         ) => Store (U.Vector (Alt f a)) where
```

**Comment by cartazio** [comment]
So it’s a bug in how the TH here is recursively lifting vector instances into store instances.  Previously vector didn’t have this sort of functor decomposition so that case wasn’t needed before ?

**Comment by cartazio** [comment]
It should be generating a 
“Store vector (f a) => store vector (Alt f a)”
Right ?

**Comment by Shimuuar** [comment]
Yep. Previously all instances for vector only had type parameter of kind `*`. Main problem is it's difficult to work with type parameters other than `*` in TH. If you have data type `Foo (f :: * → *) (a :: *)` you may need all sort of constraints in head: `Store (f Int)`, `Store (f a)`, `Store a` or any combination of them. 

I'm not sure what is right fix here.

**Comment by cartazio** [comment]
Fuzzily one approach would be to write logic specific to some known stuff.

Eg for newtypes you could take the body type of the newtype as the head
constraint?

**Comment by mgsloan** [comment]
Hey all, thanks for digging into this!

The problem is that `deriveManyStoreUnboxVector` was just assuming all type variables needed `Store (Vector a)` instances.  I resolved this by reifying the instances of `Unbox` and using that to inform the correct constraints.  So now we get

```haskell
instance Store (Vector (f (g a))) => Store (Vector (Compose f g a))
```

instead of the silly wrong thing

```
instance (Store (Vector f), Store (Vector g), Store (Vector a)) => Store (Vector (Compose f g a))
```

because the instance head is now based on substituting `Unbox ...` with `Store (Vector ...)` in the Unbox instances

```
instance Unbox (f (g a)) => Unbox (Compose f g a)
```

**Comment by mgsloan** [comment]
Fixed in version 0.7.2
